src/Raven.Server/ServerWide/CompareExchangeExpirationStorage.cs

```diff
@@ -0,0 +1,165 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Globalization;
+using Raven.Client;
+using Raven.Server.Rachis;
+using Raven.Server.ServerWide.Context;
+using Sparrow;
+using Sparrow.Binary;
+using Sparrow.Json;
+using Voron;
+using Voron.Data.Tables;
+
+namespace Raven.Server.ServerWide
+{
+    public class CompareExchangeExpirationStorage
+    {
+        public static string CompareExchangeByExpiration = "CompareExchangeByExpiration";
+
+        public static unsafe void Put(ClusterOperationContext context, Slice keySlice, long ticks)
+        {
+            var ticksBigEndian = Bits.SwapBytes(ticks);
+            using (Slice.External(context.Allocator, (byte*)&ticksBigEndian, sizeof(long), out Slice ticksSlice))
+            {
+                var tree = context.Transaction.InnerTransaction.ReadTree(CompareExchangeByExpiration);
+                tree.MultiAdd(ticksSlice, keySlice);
+            }
+        }
+
+        public static bool HasExpired(TransactionOperationContext context, long currentTicks)
+        {
+            var expirationTree = context.Transaction.InnerTransaction.ReadTree(CompareExchangeByExpiration);
+            using (var it = expirationTree.Iterate(false))
+            {
+                if (it.Seek(Slices.BeforeAllKeys) == false)
+                    return false;
+
+                var entryTicks = it.CurrentKey.CreateReader().ReadBigEndianInt64();
+                if (entryTicks > currentTicks)
+                    return false;
+            }
+
+            return true;
+        }
+
+        private static IEnumerable<(Slice keySlice, long expiredTicks, Slice ticksSlice)> GetExpiredValues(ClusterOperationContext context, long currentTicks)
+        {
+            var expirationTree = context.Transaction.InnerTransaction.ReadTree(CompareExchangeByExpiration);
+            using (var it = expirationTree.Iterate(false))
+            {
+                if (it.Seek(Slices.BeforeAllKeys) == false)
+                    yield break;
+
+                do
+                {
+                    var entryTicks = it.CurrentKey.CreateReader().ReadBigEndianInt64();
+                    if (entryTicks > currentTicks)
```

Comment:
Shouldn't this be `>=` ?

Reply:
no this is right, we should delete if the time is equal